<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>EpiTxDb objects — EpiTxDb-class • EpiTxDb</title>


<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous" />

<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script>

<!-- bootstrap-toc -->
<link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous" />
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script>

<!-- headroom.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>




<meta property="og:title" content="EpiTxDb objects — EpiTxDb-class" />
<meta property="og:description" content="The EpiTxDb class is a
AnnotationDb type container
for storing Epitranscriptomic information.
The information are typically stored on a per transcript and not as genomic
coordinates, but the EpiTxDb class is agnostic to this. In case of
genomic coordinates transcriptsBy will return modifications per
chromosome." />




<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->



  </head>

  <body data-spy="scroll" data-target="#toc">
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">EpiTxDb</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.3.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/EpiTxDb.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/EpiTxDb-creation.html">EpiTxDb: creating an EpiTxDb object</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/FelixErnst/EpiTxDb/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>EpiTxDb objects</h1>
    <small class="dont-index">Source: <a href='https://github.com/FelixErnst/EpiTxDb/blob/master/R/EpiTxDb-class.R'><code>R/EpiTxDb-class.R</code></a></small>
    <div class="hidden name"><code>EpiTxDb-class.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>The <code>EpiTxDb</code> class is a
<code><a href='https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html'>AnnotationDb</a></code> type container
for storing Epitranscriptomic information.</p>
<p>The information are typically stored on a per transcript and not as genomic
coordinates, but the <code>EpiTxDb</code> class is agnostic to this. In case of
genomic coordinates <code>transcriptsBy</code> will return modifications per
chromosome.</p>
    </div>

    <pre class="usage"><span class='co'># S4 method for EpiTxDb</span>
<span class='fu'>organism</span><span class='op'>(</span><span class='va'>object</span><span class='op'>)</span>

<span class='co'># S4 method for EpiTxDb</span>
<span class='fu'>seqinfo</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span>

<span class='co'># S4 method for EpiTxDb</span>
<span class='fu'>seqlevels</span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span>

<span class='co'># S4 method for EpiTxDb</span>
<span class='fu'><a href='https://rdrr.io/r/base/list.html'>as.list</a></span><span class='op'>(</span><span class='va'>x</span><span class='op'>)</span></pre>

    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>x, object</th>
      <td><p>a <code>EpiTxDb</code> object</p></td>
    </tr>
    </table>

    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>For</p><ul>
<li><p><code>organism()</code> and <code>seqlevels()</code> a <code>character</code>
    vector</p></li>
<li><p><code>seqinfo()</code> a
    <code><a href='https://rdrr.io/pkg/GenomeInfoDb/man/Seqinfo-class.html'>Seqinfo</a></code> object</p></li>
<li><p><code><a href='https://rdrr.io/r/base/list.html'>as.list()</a></code> a <code>list</code></p></li>
</ul>

    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'>
<ul>
<li><p><code><a href='makeEpiTxDbFromGRanges.html'>makeEpiTxDbFromGRanges</a></code> for
      creating a <code>EpiTxDb</code> object from a
      <code><a href='https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html'>GRanges</a></code> object and it's
       metadata columns</p></li>
<li><p><code><a href='makeEpiTxDbFromRMBase.html'>makeEpiTxDbFromRMBase</a></code> for
      creating a <code>EpiTxDb</code> object from RMBase online resources</p></li>
<li><p><code><a href='makeEpiTxDbFromtRNAdb.html'>makeEpiTxDbFromtRNAdb</a></code> for
      creating a <code>EpiTxDb</code> object from tRNAdb online resources</p></li>
<li><p><code><a href='makeEpiTxDb.html'>makeEpiTxDb</a></code> for creating a
      <code>EpiTxDb</code> object from <code>data.frames</code></p></li>
<li><p><code><a href='modifications.html'>modifications</a></code>,
      <code><a href='modifications.html'>modificationsBy</a></code> for getting
      epitranscriptomic modification locations</p></li>
<li><p><code><a href='select.html'>select</a></code> for using the default interface of
      <code><a href='https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html'>AnnotationDb</a></code> objects.</p></li>
<li><p><code><a href='shiftTranscriptToGenomic.html'>shiftGenomicToTranscript</a></code>
      and <code><a href='shiftTranscriptToGenomic.html'>shiftTranscriptToGenomic</a></code>
      for transfering genomic to transcript coordinates and back again.</p></li>
</ul>
</div>

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='va'>etdb_file</span> <span class='op'>&lt;-</span> <span class='fu'><a href='https://rdrr.io/r/base/system.file.html'>system.file</a></span><span class='op'>(</span><span class='st'>"extdata"</span>, <span class='st'>"EpiTxDb.Hs.hg38.snoRNAdb.sqlite"</span>,
                        package<span class='op'>=</span><span class='st'>"EpiTxDb"</span><span class='op'>)</span>
<span class='va'>etdb</span> <span class='op'>&lt;-</span> <span class='fu'>loadDb</span><span class='op'>(</span><span class='va'>etdb_file</span><span class='op'>)</span>
<span class='va'>etdb</span>
</div><div class='output co'>#&gt; EpiTxDb object:
#&gt; # Db type: EpiTxDb
#&gt; # Supporting package: EpiTxDb
#&gt; # Data source: snoRNAdb
#&gt; # Organism: Homo sapiens
#&gt; # Genome: hg38
#&gt; # Coordinates: per Transcript
#&gt; # Nb of modifications: 235
#&gt; # Db created by: EpiTxDb package from Bioconductor
#&gt; # Creation time: 2020-02-26 10:34:30 +0100 (Wed, 26 Feb 2020)
#&gt; # EpiTxDb version at creation time: 0.99.0
#&gt; # RSQLite version at creation time: 2.2.0
#&gt; # DBSCHEMAVERSION: 1.0</div><div class='input'>
<span class='co'># general methods</span>
<span class='fu'>seqinfo</span><span class='op'>(</span><span class='va'>etdb</span><span class='op'>)</span> <span class='co'>#</span>
</div><div class='output co'>#&gt; Seqinfo object with 9 sequences from hg38 genome; no seqlengths:
#&gt;   seqnames  seqlengths isCircular genome
#&gt;   NR_003285         NA         NA   hg38
#&gt;   NR_003286         NA         NA   hg38
#&gt;   NR_003287         NA         NA   hg38
#&gt;   NR_004430         NA         NA   hg38
#&gt;   NR_002716         NA         NA   hg38
#&gt;   NR_003925         NA         NA   hg38
#&gt;   NR_002756         NA         NA   hg38
#&gt;   NR_004394         NA         NA   hg38
#&gt;   NR_029422         NA         NA   hg38</div><div class='input'><span class='fu'>seqlevels</span><span class='op'>(</span><span class='va'>etdb</span><span class='op'>)</span> <span class='co'># easy access to all transcript names</span>
</div><div class='output co'>#&gt; [1] "NR_003285" "NR_003286" "NR_003287" "NR_004430" "NR_002716" "NR_003925"
#&gt; [7] "NR_002756" "NR_004394" "NR_029422"</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top">
      <h2 data-toc-skip>Contents</h2>
    </nav>
  </div>
</div>


      <footer>
      <div class="copyright">
  <p>Developed by Felix G.M. Ernst.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
   </div>

  


  </body>
</html>


