<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Creating a EpiTxDb from user supplied annotations as
data.frames — makeEpiTxDb • EpiTxDb</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Creating a EpiTxDb from user supplied annotations as
data.frames — makeEpiTxDb"><meta property="og:description" content="makeEpiTxDb is a low-level constructor for creating a
EpiTxDb object from user supplied annotations.
This functions typically will not be used by regular users."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">EpiTxDb</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.15.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/EpiTxDb.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/EpiTxDb-creation.html">EpiTxDb: creating an EpiTxDb object</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/FelixErnst/EpiTxDb/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Creating a <code>EpiTxDb</code> from user supplied annotations as
<code>data.frame</code>s</h1>
    <small class="dont-index">Source: <a href="https://github.com/FelixErnst/EpiTxDb/blob/HEAD/R/makeEpiTxDb.R" class="external-link"><code>R/makeEpiTxDb.R</code></a></small>
    <div class="hidden name"><code>makeEpiTxDb.Rd</code></div>
    </div>

    <div class="ref-description">
    <p><code>makeEpiTxDb</code> is a low-level constructor for creating a
<code><a href="EpiTxDb-class.html">EpiTxDb</a></code> object from user supplied annotations.</p>
<p>This functions typically will not be used by regular users.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">makeEpiTxDb</span><span class="op">(</span></span>
<span>  <span class="va">modifications</span>,</span>
<span>  reactions <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  specifiers <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  references <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  metadata <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  reassign.ids <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>modifications</dt>
<dd><p>A <code>data.frame</code> containg the following columns:</p><ul><li><p><code>mod_id</code>: a unique <code>integer</code> value per modification.</p></li>
<li><p><code>mod_type</code>: the modification type as a <code>character</code> or
  <code>factor</code> value. Must be a value from
  <code>shortName(ModRNAString())</code>.</p></li>
<li><p><code>mod_name</code>: a <code>character</code> or <code>factor</code> name for the
  specific modification</p></li>
<li><p><code>mod_start</code>: the start position for the modification as
  <code>integer</code> value. Usually <code>mod_start = mod_end</code></p></li>
<li><p><code>mod_end</code>: the end position for the modification as
  <code>integer</code> value. Usually <code>mod_start = mod_end</code></p></li>
<li><p><code>mod_strand</code>: the strand information for the modificaion as a
  <code>character</code> or <code>factor</code>.</p></li>
<li><p><code>sn_id</code>: a <code>integer</code> value per unique sequence</p></li>
<li><p><code>sn_name</code>: a <code>character</code> or <code>factor</code> as sequence
  name, e.g a chromosome or a transcript identifier like <code>chr1</code>.</p></li>
</ul><p>The first six are mandatory, whereas one of the last two has to be set.
<code>sn_id</code> will be generated from <code>sn_name</code>, if <code>sn_id</code> is not
set.</p></dd>


<dt>reactions</dt>
<dd><p>An optional <code>data.frame</code> containg the following
columns:</p><ul><li><p><code>mod_id</code>: a <code>integer</code> value per modification and the
  link to the <code>modification</code> <code>data.frame</code>.</p></li>
<li><p><code>rx_genename</code>: a <code>character</code> or <code>factor</code> referencing
  a genename for the enzyme incorporating the modification.</p></li>
<li><p><code>rx_rank</code>: a <code>integer</code> for sorting enzyme reactions, if
  multiple enzymes are involved in the modification's
  incorporation/maintenance.</p></li>
<li><p><code>rx_ensembl</code>: a <code>character</code> or <code>factor</code> with an
  ensembl identifier for the genename of the enzyme.</p></li>
<li><p><code>rx_ensembltrans</code>: a <code>character</code> or <code>factor</code> with an
  ensembl identifier for the transcript being translated into the enzyme.</p></li>
<li><p><code>rx_entrezid</code>: a <code>character</code> or <code>factor</code> with an
  entrezid for the genename of the enzyme.</p></li>
</ul><p>(default: <code>reactions = NULL</code>)</p></dd>


<dt>specifiers</dt>
<dd><p>An optional <code>data.frame</code> containg the following
columns:</p><ul><li><p><code>mod_id</code>: a <code>integer</code> value per modification and the
  link to the <code>modification</code> <code>data.frame</code>.</p></li>
<li><p><code>spec_type</code>: a <code>character</code> or <code>factor</code>
  referencing a type of specifier, e.g. <code>snoRNA</code>. Not checked for
  validity.</p></li>
<li><p><code>spec_genename</code>: a <code>character</code> or <code>factor</code>
  referencing a genename for the specifier directing an enzyme to the
  specific location for the modification to be incorporated.</p></li>
<li><p><code>spec_ensembl</code>: a <code>character</code> or <code>factor</code> with an
  ensembl identifier for the genename of the specifier.</p></li>
<li><p><code>spec_ensembltrans</code>: a <code>character</code> or <code>factor</code> with an
  ensembl identifier for the transcript being translated into the specifier.</p></li>
<li><p><code>spec_entrezid</code>: a <code>character</code> or <code>factor</code> with an
  entrezid for the genename of the specifier.</p></li>
</ul><p>(default: <code>specifiers = NULL</code>)</p></dd>


<dt>references</dt>
<dd><p>An optional <code>data.frame</code> containg the following
columns:</p><ul><li><p><code>mod_id</code>: a <code>integer</code> value per modification and the
  link to the <code>modification</code> <code>data.frame</code>.</p></li>
<li><p><code>ref_type</code>: a <code>character</code> or <code>factor</code> with a
  reference type, e.g. <code>PMID</code>. Is not checked for validity.</p></li>
<li><p><code>ref</code>: a <code>character</code> or <code>factor</code> with a reference
  value, e.g. a specific pubmed id or an journal article. Is not checked
  for validity.</p></li>
</ul><p>(default: <code>references = NULL</code>)</p></dd>


<dt>metadata</dt>
<dd><p>An optional <code>data.frame</code> containg the following columns:</p><ul><li><p><code>name</code>: a <code>character</code> value used as name</p></li>
<li><p><code>value</code>: a <code>character</code> value</p></li>
</ul><p>This dataframe will be returned
by <code>metadata()</code> (default: <code>metadata = NULL</code>)</p></dd>


<dt>reassign.ids</dt>
<dd><p><code>TRUE</code> or <code>FALSE</code> Controls how internal
<code>mod_id</code>s should be assigned. If <code>reassign.ids</code> is <code>FALSE</code>
(the default) and if the ids are supplied, then they are used as the
internal ids, otherwise the internal ids are assigned in a way that is
compatible with the order defined by ordering the modifications first by
chromosome, then by strand, then by start, and finally by end.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>a <code>EpiTxDb</code> object.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index">
<ul><li><p><code><a href="makeEpiTxDbFromGRanges.html">makeEpiTxDbFromGRanges</a></code> for
    creating a <code>EpiTxDb</code> object from a
  <code><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></code> object and it's metadata
    columns</p></li>
<li><p><code><a href="makeEpiTxDbFromRMBase.html">makeEpiTxDbFromRMBase</a></code> for
    creating a <code>EpiTxDb</code> object from RMBase online resources</p></li>
<li><p><code><a href="makeEpiTxDbFromtRNAdb.html">makeEpiTxDbFromtRNAdb</a></code> for
    creating a <code>EpiTxDb</code> object from tRNAdb online resources</p></li>
<li><p><code><a href="https://rdrr.io/pkg/Biostrings/man/XString-class.html" class="external-link">shortName</a></code> and
  <code><a href="https://rdrr.io/pkg/Modstrings/man/ModRNAString.html" class="external-link">ModRNAString</a></code> for information on
  <code>ModRNAString</code> objects.</p></li>
</ul></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">mod</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="st">"mod_id"</span> <span class="op">=</span> <span class="fl">1L</span>,</span></span>
<span class="r-in"><span>                  <span class="st">"mod_type"</span> <span class="op">=</span> <span class="st">"m1A"</span>,</span></span>
<span class="r-in"><span>                  <span class="st">"mod_name"</span> <span class="op">=</span> <span class="st">"m1A_1"</span>,</span></span>
<span class="r-in"><span>                  <span class="st">"mod_start"</span> <span class="op">=</span> <span class="fl">1L</span>,</span></span>
<span class="r-in"><span>                  <span class="st">"mod_end"</span> <span class="op">=</span> <span class="fl">1L</span>,</span></span>
<span class="r-in"><span>                  <span class="st">"mod_strand"</span> <span class="op">=</span> <span class="st">"+"</span>,</span></span>
<span class="r-in"><span>                  <span class="st">"sn_id"</span> <span class="op">=</span> <span class="fl">1L</span>,</span></span>
<span class="r-in"><span>                  <span class="st">"sn_name"</span> <span class="op">=</span> <span class="st">"test"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">rx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>mod_id <span class="op">=</span> <span class="fl">1L</span>,</span></span>
<span class="r-in"><span>                 rx_genename <span class="op">=</span> <span class="st">"test"</span>,</span></span>
<span class="r-in"><span>                 rx_rank <span class="op">=</span> <span class="fl">1L</span>,</span></span>
<span class="r-in"><span>                 rx_ensembl <span class="op">=</span> <span class="st">"test"</span>,</span></span>
<span class="r-in"><span>                 rx_ensembltrans <span class="op">=</span> <span class="st">"test"</span>,</span></span>
<span class="r-in"><span>                 rx_entrezid <span class="op">=</span> <span class="st">"test"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">spec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>mod_id <span class="op">=</span> <span class="fl">1L</span>,</span></span>
<span class="r-in"><span>                   spec_type <span class="op">=</span> <span class="st">"test"</span>,</span></span>
<span class="r-in"><span>                   spec_genename <span class="op">=</span> <span class="st">"test"</span>,</span></span>
<span class="r-in"><span>                   spec_ensembl <span class="op">=</span> <span class="st">"test"</span>,</span></span>
<span class="r-in"><span>                   spec_ensembltrans <span class="op">=</span> <span class="st">"test"</span>,</span></span>
<span class="r-in"><span>                   spec_entrezid <span class="op">=</span> <span class="st">"test"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">ref</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>mod_id <span class="op">=</span> <span class="fl">1L</span>,</span></span>
<span class="r-in"><span>                  ref_type <span class="op">=</span> <span class="st">"test"</span>,</span></span>
<span class="r-in"><span>                  ref <span class="op">=</span> <span class="st">"test"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">etdb</span> <span class="op">&lt;-</span> <span class="fu">makeEpiTxDb</span><span class="op">(</span><span class="va">mod</span>,<span class="va">rx</span>,<span class="va">spec</span>,<span class="va">ref</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Creating EpiTxDb object ... </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> done</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Felix G.M. Ernst.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

