<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>EpiTxDb: creating an EpiTxDb object • EpiTxDb</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="EpiTxDb: creating an EpiTxDb object">
<meta property="og:description" content="EpiTxDb">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">EpiTxDb</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.15.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/EpiTxDb.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/EpiTxDb-creation.html">EpiTxDb: creating an EpiTxDb object</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/FelixErnst/EpiTxDb/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>EpiTxDb: creating an EpiTxDb object</h1>
                        <h4 data-toc-skip class="author">Felix G.M.
Ernst</h4>
            
            <h4 data-toc-skip class="date">2024-03-18</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/FelixErnst/EpiTxDb/blob/HEAD/vignettes/EpiTxDb-creation.Rmd" class="external-link"><code>vignettes/EpiTxDb-creation.Rmd</code></a></small>
      <div class="hidden name"><code>EpiTxDb-creation.Rmd</code></div>

    </div>

    
    
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script><script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script><div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>To create an <code>EpiTxDb</code> object a number of different
functions are available. The most univeral functions are
<code>makeEpiTxDb</code> and <code>makeEpiTxDbFromGRanges</code>.
<code>makeEpiTxDb</code> uses four <code>data.frame</code>s as input,
whereas <code>makeEpiTxDbFromGRanges</code> is a wrapper for information
available as a <code>GRanges</code> object.</p>
<p>The other functions are <code>makeEpiTxDbFromRMBase</code> and
<code>makeEpiTxDbFromtRNAdb</code>, which are aimed to make data
available from the RMBase v2.0 database <a href="#References"><span class="citation">(Xuan et al. 2017; Sun et al. 2015)</span></a> or the
tRNAdb <a href="#References"><span class="citation">(Jühling et al.
2009; Sprinzl and Vassilenko 2005)</span></a>. However, before creating
your <code>EpiTxDb</code> objects, have a look at the already available
resources for <code>H. sapiens</code>, <code>M. musculus</code> and
<code>S. cerevisiae.</code></p>
<p>Additional metadata can be provided as separate
<code>data.frame</code> for all functions. The <code>data.frame</code>
must have two columns <code>name</code> and <code>value</code>.</p>
<pre><code><span><span class="co">## Warning: replacing previous import 'txdbmaker::makeTxDbFromBiomart' by</span></span>
<span><span class="co">## 'GenomicFeatures::makeTxDbFromBiomart' when loading 'EpiTxDb'</span></span></code></pre>
<pre><code><span><span class="co">## Warning: replacing previous import 'txdbmaker::supportedUCSCFeatureDbTables' by</span></span>
<span><span class="co">## 'GenomicFeatures::supportedUCSCFeatureDbTables' when loading 'EpiTxDb'</span></span></code></pre>
<pre><code><span><span class="co">## Warning: replacing previous import 'txdbmaker::getChromInfoFromBiomart' by</span></span>
<span><span class="co">## 'GenomicFeatures::getChromInfoFromBiomart' when loading 'EpiTxDb'</span></span></code></pre>
<pre><code><span><span class="co">## Warning: replacing previous import 'txdbmaker::makeTxDbFromGRanges' by</span></span>
<span><span class="co">## 'GenomicFeatures::makeTxDbFromGRanges' when loading 'EpiTxDb'</span></span></code></pre>
<pre><code><span><span class="co">## Warning: replacing previous import 'txdbmaker::makeTxDb' by</span></span>
<span><span class="co">## 'GenomicFeatures::makeTxDb' when loading 'EpiTxDb'</span></span></code></pre>
<pre><code><span><span class="co">## Warning: replacing previous import 'txdbmaker::supportedMiRBaseBuildValues' by</span></span>
<span><span class="co">## 'GenomicFeatures::supportedMiRBaseBuildValues' when loading 'EpiTxDb'</span></span></code></pre>
<pre><code><span><span class="co">## Warning: replacing previous import 'txdbmaker::makeTxDbPackage' by</span></span>
<span><span class="co">## 'GenomicFeatures::makeTxDbPackage' when loading 'EpiTxDb'</span></span></code></pre>
<pre><code><span><span class="co">## Warning: replacing previous import 'txdbmaker::makeTxDbFromEnsembl' by</span></span>
<span><span class="co">## 'GenomicFeatures::makeTxDbFromEnsembl' when loading 'EpiTxDb'</span></span></code></pre>
<pre><code><span><span class="co">## Warning: replacing previous import 'txdbmaker::makeFeatureDbFromUCSC' by</span></span>
<span><span class="co">## 'GenomicFeatures::makeFeatureDbFromUCSC' when loading 'EpiTxDb'</span></span></code></pre>
<pre><code><span><span class="co">## Warning: replacing previous import 'txdbmaker::makePackageName' by</span></span>
<span><span class="co">## 'GenomicFeatures::makePackageName' when loading 'EpiTxDb'</span></span></code></pre>
<pre><code><span><span class="co">## Warning: replacing previous import 'txdbmaker::UCSCFeatureDbTableSchema' by</span></span>
<span><span class="co">## 'GenomicFeatures::UCSCFeatureDbTableSchema' when loading 'EpiTxDb'</span></span></code></pre>
<pre><code><span><span class="co">## Warning: replacing previous import 'txdbmaker::makeFDbPackageFromUCSC' by</span></span>
<span><span class="co">## 'GenomicFeatures::makeFDbPackageFromUCSC' when loading 'EpiTxDb'</span></span></code></pre>
<pre><code><span><span class="co">## Warning: replacing previous import 'txdbmaker::makeTxDbFromUCSC' by</span></span>
<span><span class="co">## 'GenomicFeatures::makeTxDbFromUCSC' when loading 'EpiTxDb'</span></span></code></pre>
<pre><code><span><span class="co">## Warning: replacing previous import 'txdbmaker::browseUCSCtrack' by</span></span>
<span><span class="co">## 'GenomicFeatures::browseUCSCtrack' when loading 'EpiTxDb'</span></span></code></pre>
<pre><code><span><span class="co">## Warning: replacing previous import 'txdbmaker::supportedUCSCtables' by</span></span>
<span><span class="co">## 'GenomicFeatures::supportedUCSCtables' when loading 'EpiTxDb'</span></span></code></pre>
<pre><code><span><span class="co">## Warning: replacing previous import 'txdbmaker::makeTxDbPackageFromUCSC' by</span></span>
<span><span class="co">## 'GenomicFeatures::makeTxDbPackageFromUCSC' when loading 'EpiTxDb'</span></span></code></pre>
<pre><code><span><span class="co">## Warning: replacing previous import 'txdbmaker::supportedUCSCFeatureDbTracks' by</span></span>
<span><span class="co">## 'GenomicFeatures::supportedUCSCFeatureDbTracks' when loading 'EpiTxDb'</span></span></code></pre>
<pre><code><span><span class="co">## Warning: replacing previous import 'txdbmaker::makeTxDbPackageFromBiomart' by</span></span>
<span><span class="co">## 'GenomicFeatures::makeTxDbPackageFromBiomart' when loading 'EpiTxDb'</span></span></code></pre>
<pre><code><span><span class="co">## Warning: replacing previous import 'txdbmaker::makeTxDbFromGFF' by</span></span>
<span><span class="co">## 'GenomicFeatures::makeTxDbFromGFF' when loading 'EpiTxDb'</span></span></code></pre>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges" class="external-link">GenomicRanges</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/FelixErnst/EpiTxDb" class="external-link">EpiTxDb</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="makeepitxdb-and-makeepitxdbfromgranges">
<code>makeEpiTxDb</code> and
<code>makeEpiTxDbFromGRanges</code><a class="anchor" aria-label="anchor" href="#makeepitxdb-and-makeepitxdbfromgranges"></a>
</h2>
<p>The creation of an etdb object is quite easy starting with a
<code>GRanges</code> object.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span>seqnames <span class="op">=</span> <span class="st">"test"</span>,</span>
<span>              ranges <span class="op">=</span> <span class="fu">IRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span>,</span>
<span>              strand <span class="op">=</span> <span class="st">"+"</span>,</span>
<span>              <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html" class="external-link">DataFrame</a></span><span class="op">(</span>mod_id <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>                        mod_type <span class="op">=</span> <span class="st">"Am"</span>,</span>
<span>                        mod_name <span class="op">=</span> <span class="st">"Am_1"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">etdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/makeEpiTxDbFromGRanges.html">makeEpiTxDbFromGRanges</a></span><span class="op">(</span><span class="va">gr</span>, metadata <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>name <span class="op">=</span> <span class="st">"test"</span>,</span>
<span>                                                         value <span class="op">=</span> <span class="st">"Yes"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning in as.data.frame.factor(value, row.names = row.names, optional =</span></span>
<span><span class="co">## optional, : Direct call of 'as.data.frame.factor()' is deprecated.  Use</span></span>
<span><span class="co">## 'as.data.frame.vector()' or 'as.data.frame()' instead</span></span>
<span></span>
<span><span class="co">## Warning in as.data.frame.factor(value, row.names = row.names, optional =</span></span>
<span><span class="co">## optional, : Direct call of 'as.data.frame.factor()' is deprecated.  Use</span></span>
<span><span class="co">## 'as.data.frame.vector()' or 'as.data.frame()' instead</span></span></code></pre>
<pre><code><span><span class="co">## Creating EpiTxDb object ... done</span></span></code></pre>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">etdb</span></span></code></pre></div>
<pre><code><span><span class="co">## EpiTxDb object:</span></span>
<span><span class="co">## # Db type: EpiTxDb</span></span>
<span><span class="co">## # Supporting package: EpiTxDb</span></span>
<span><span class="co">## # test: Yes</span></span>
<span><span class="co">## # Nb of modifications: 1</span></span>
<span><span class="co">## # Db created by: EpiTxDb package from Bioconductor</span></span>
<span><span class="co">## # Creation time: 2024-03-18 21:00:47 +0000 (Mon, 18 Mar 2024)</span></span>
<span><span class="co">## # EpiTxDb version at creation time: 1.15.2</span></span>
<span><span class="co">## # RSQLite version at creation time: 2.3.5</span></span>
<span><span class="co">## # DBSCHEMAVERSION: 1.0</span></span></code></pre>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Annotated-class.html" class="external-link">metadata</a></span><span class="op">(</span><span class="va">etdb</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                               name                                        value</span></span>
<span><span class="co">## 1                          Db type                                      EpiTxDb</span></span>
<span><span class="co">## 2               Supporting package                                      EpiTxDb</span></span>
<span><span class="co">## 3                             test                                          Yes</span></span>
<span><span class="co">## 4              Nb of modifications                                            1</span></span>
<span><span class="co">## 5                    Db created by            EpiTxDb package from Bioconductor</span></span>
<span><span class="co">## 6                    Creation time 2024-03-18 21:00:47 +0000 (Mon, 18 Mar 2024)</span></span>
<span><span class="co">## 7 EpiTxDb version at creation time                                       1.15.2</span></span>
<span><span class="co">## 8 RSQLite version at creation time                                        2.3.5</span></span>
<span><span class="co">## 9                  DBSCHEMAVERSION                                          1.0</span></span></code></pre>
<p>Additional data can be provided via the metadata columns of the
<code>GRanges</code> object. For supported columns have a look at
<code><a href="../reference/makeEpiTxDb.html">?makeEpiTxDb</a></code> or <code><a href="../reference/makeEpiTxDbFromGRanges.html">?makeEpiTxDbFromGRanges</a></code>.</p>
</div>
<div class="section level2">
<h2 id="makeepitxdbfromtrnadb">
<code>makeEpiTxDbFromtRNAdb</code><a class="anchor" aria-label="anchor" href="#makeepitxdbfromtrnadb"></a>
</h2>
<p>The information of the tRNAdb can be accessed via the
<code>tRNAdbImport</code> package using the RNA database. As a result a
<code>ModRNAStringSet</code> object is returned from which the
modifications can be extracted using <code><a href="https://rdrr.io/pkg/Modstrings/man/separate.html" class="external-link">separate()</a></code>.</p>
<p>The only input require is a valid organism name returned by
<code><a href="../reference/makeEpiTxDbFromtRNAdb.html">listAvailableOrganismsFromtRNAdb()</a></code>.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Currently not run since the server is not available</span></span>
<span><span class="va">etdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/makeEpiTxDbFromtRNAdb.html">makeEpiTxDbFromtRNAdb</a></span><span class="op">(</span><span class="st">"Saccharomyces cerevisiae"</span><span class="op">)</span></span>
<span><span class="va">etdb</span></span></code></pre></div>
<p>For additional information have a look at
<code><a href="../reference/makeEpiTxDbFromtRNAdb.html">?makeEpiTxDbFromtRNAdb</a></code>. The result returned from the tRNAdb
is also available as <code>GRanges</code> object, if
<code><a href="../reference/makeEpiTxDbFromtRNAdb.html">gettRNAdbDataAsGRanges()</a></code> is used.</p>
</div>
<div class="section level2">
<h2 id="makeepitxdbfromrmbase">
<code>makeEpiTxDbFromRMBase</code><a class="anchor" aria-label="anchor" href="#makeepitxdbfromrmbase"></a>
</h2>
<p>Analogous to the example above <code><a href="../reference/makeEpiTxDbFromRMBase.html">makeEpiTxDbFromRMBase()</a></code>
will download the data from the RMBase v2.0. Three inputs are required,
<code>organism</code>, <code>genome</code> and <code>modtype</code>,
which have to valid bia the functions
<code><a href="../reference/makeEpiTxDbFromRMBase.html">listAvailableOrganismsFromRMBase()</a></code>,
<code>.listAvailableGenomesFromRMBase()</code> and
<code>listAvailableModFromRMBase</code>.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">etdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/makeEpiTxDbFromRMBase.html">makeEpiTxDbFromRMBase</a></span><span class="op">(</span>organism <span class="op">=</span> <span class="st">"Saccharomyces cerevisiae"</span>,</span>
<span>                              genome <span class="op">=</span> <span class="st">"sacCer3"</span>,</span>
<span>                              modtype <span class="op">=</span> <span class="st">"m1A"</span><span class="op">)</span></span></code></pre></div>
<p>Internally, the files are cached using the <code>BiocFileCache</code>
package and passed to <code><a href="../reference/makeEpiTxDbFromRMBase.html">makeEpiTxDbFromRMBaseFiles()</a></code>, which
can also be used with locally stored files. The resuls for creating the
<code>EpiTxDb</code> class are processed from these files via the
<code><a href="../reference/makeEpiTxDbFromRMBase.html">getRMBaseDataAsGRanges()</a></code> function.</p>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R Under development (unstable) (2024-03-13 r86113)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 22.04.4 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] EpiTxDb_1.15.2       Modstrings_1.19.0    Biostrings_2.71.4   </span></span>
<span><span class="co">##  [4] XVector_0.43.1       AnnotationDbi_1.65.2 Biobase_2.63.0      </span></span>
<span><span class="co">##  [7] GenomicRanges_1.55.3 GenomeInfoDb_1.39.9  IRanges_2.37.1      </span></span>
<span><span class="co">## [10] S4Vectors_0.41.4     BiocGenerics_0.49.1  BiocStyle_2.31.0    </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] DBI_1.2.2                   bitops_1.0-7               </span></span>
<span><span class="co">##  [3] httr2_1.0.0                 tRNAdbImport_1.21.1        </span></span>
<span><span class="co">##  [5] biomaRt_2.59.1              rlang_1.1.3                </span></span>
<span><span class="co">##  [7] magrittr_2.0.3              matrixStats_1.2.0          </span></span>
<span><span class="co">##  [9] compiler_4.4.0              RSQLite_2.3.5              </span></span>
<span><span class="co">## [11] GenomicFeatures_1.55.4      png_0.1-8                  </span></span>
<span><span class="co">## [13] systemfonts_1.0.6           vctrs_0.6.5                </span></span>
<span><span class="co">## [15] txdbmaker_0.99.6            stringr_1.5.1              </span></span>
<span><span class="co">## [17] pkgconfig_2.0.3             crayon_1.5.2               </span></span>
<span><span class="co">## [19] fastmap_1.1.1               dbplyr_2.4.0               </span></span>
<span><span class="co">## [21] utf8_1.2.4                  Rsamtools_2.19.3           </span></span>
<span><span class="co">## [23] rmarkdown_2.26              ragg_1.3.0                 </span></span>
<span><span class="co">## [25] purrr_1.0.2                 bit_4.0.5                  </span></span>
<span><span class="co">## [27] xfun_0.42                   zlibbioc_1.49.3            </span></span>
<span><span class="co">## [29] cachem_1.0.8                Structstrings_1.19.0       </span></span>
<span><span class="co">## [31] jsonlite_1.8.8              progress_1.2.3             </span></span>
<span><span class="co">## [33] blob_1.2.4                  DelayedArray_0.29.9        </span></span>
<span><span class="co">## [35] BiocParallel_1.37.1         parallel_4.4.0             </span></span>
<span><span class="co">## [37] prettyunits_1.2.0           R6_2.5.1                   </span></span>
<span><span class="co">## [39] bslib_0.6.1                 stringi_1.8.3              </span></span>
<span><span class="co">## [41] rtracklayer_1.63.1          jquerylib_0.1.4            </span></span>
<span><span class="co">## [43] bookdown_0.38               SummarizedExperiment_1.33.3</span></span>
<span><span class="co">## [45] knitr_1.45                  Matrix_1.6-5               </span></span>
<span><span class="co">## [47] tidyselect_1.2.1            abind_1.4-5                </span></span>
<span><span class="co">## [49] yaml_2.3.8                  codetools_0.2-19           </span></span>
<span><span class="co">## [51] curl_5.2.1                  tRNA_1.21.1                </span></span>
<span><span class="co">## [53] lattice_0.22-5              tibble_3.2.1               </span></span>
<span><span class="co">## [55] KEGGREST_1.43.0             evaluate_0.23              </span></span>
<span><span class="co">## [57] desc_1.4.3                  BiocFileCache_2.11.1       </span></span>
<span><span class="co">## [59] xml2_1.3.6                  pillar_1.9.0               </span></span>
<span><span class="co">## [61] BiocManager_1.30.22         filelock_1.0.3             </span></span>
<span><span class="co">## [63] MatrixGenerics_1.15.0       generics_0.1.3             </span></span>
<span><span class="co">## [65] RCurl_1.98-1.14             hms_1.1.3                  </span></span>
<span><span class="co">## [67] ggplot2_3.5.0               munsell_0.5.0              </span></span>
<span><span class="co">## [69] scales_1.3.0                glue_1.7.0                 </span></span>
<span><span class="co">## [71] tools_4.4.0                 BiocIO_1.13.0              </span></span>
<span><span class="co">## [73] GenomicAlignments_1.39.4    fs_1.6.3                   </span></span>
<span><span class="co">## [75] XML_3.99-0.16.1             grid_4.4.0                 </span></span>
<span><span class="co">## [77] colorspace_2.1-0            GenomeInfoDbData_1.2.11    </span></span>
<span><span class="co">## [79] restfulr_0.0.15             cli_3.6.2                  </span></span>
<span><span class="co">## [81] rappdirs_0.3.3              textshaping_0.3.7          </span></span>
<span><span class="co">## [83] fansi_1.0.6                 S4Arrays_1.3.6             </span></span>
<span><span class="co">## [85] dplyr_1.1.4                 gtable_0.3.4               </span></span>
<span><span class="co">## [87] sass_0.4.9                  digest_0.6.35              </span></span>
<span><span class="co">## [89] SparseArray_1.3.4           rjson_0.2.21               </span></span>
<span><span class="co">## [91] memoise_2.0.1               htmltools_0.5.7            </span></span>
<span><span class="co">## [93] pkgdown_2.0.7               lifecycle_1.0.4            </span></span>
<span><span class="co">## [95] httr_1.4.7                  bit64_4.0.5</span></span></code></pre>
<p><a name="References"></a></p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Juehling.2009" class="csl-entry">
Jühling, Frank, Mario Mörl, Roland K. Hartmann, Mathias Sprinzl, Peter
F. Stadler, and Joern Pütz. 2009. <span>“tRNAdb 2009: Compilation of
tRNA Sequences and tRNA Genes.”</span> <em>Nucleic Acids Research</em>
37 (suppl_1): D159–62. <a href="https://doi.org/10.1093/nar/gkn772" class="external-link">https://doi.org/10.1093/nar/gkn772</a>.
</div>
<div id="ref-Sprinzl.2005" class="csl-entry">
Sprinzl, Mathias, and Konstantin S. Vassilenko. 2005. <span>“Compilation
of tRNA Sequences and Sequences of tRNA Genes.”</span> <em>Nucleic Acids
Research</em> 33 (suppl_1): D139–40. <a href="https://doi.org/10.1093/nar/gki012" class="external-link">https://doi.org/10.1093/nar/gki012</a>.
</div>
<div id="ref-Sun.2015" class="csl-entry">
Sun, Wen-Ju, Jun-Hao Li, Shun Liu, Jie Wu, Hui Zhou, Liang-Hu Qu, and
Jian-Hua Yang. 2015. <span>“<span class="nocase">RMBase: a resource for
decoding the landscape of RNA modifications from high-throughput
sequencing data</span>.”</span> <em>Nucleic Acids Research</em> 44 (D1):
D259–65. <a href="https://doi.org/10.1093/nar/gkv1036" class="external-link">https://doi.org/10.1093/nar/gkv1036</a>.
</div>
<div id="ref-Xuan.2017" class="csl-entry">
Xuan, Jia-Jia, Wen-Ju Sun, Peng-Hui Lin, Ke-Ren Zhou, Shun Liu,
Ling-Ling Zheng, Liang-Hu Qu, and Jian-Hua Yang. 2017. <span>“<span class="nocase">RMBase v2.0: deciphering the map of RNA modifications
from epitranscriptome sequencing data</span>.”</span> <em>Nucleic Acids
Research</em> 46 (D1): D327–34. <a href="https://doi.org/10.1093/nar/gkx934" class="external-link">https://doi.org/10.1093/nar/gkx934</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Felix G.M. Ernst.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
