<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>EpiTxDb: Storing and accessing epitranscriptomic information using the AnnotationDbi interface • EpiTxDb</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="EpiTxDb: Storing and accessing epitranscriptomic information using the AnnotationDbi interface">
<meta property="og:description" content="EpiTxDb">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">EpiTxDb</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.3.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/EpiTxDb.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/EpiTxDb-creation.html">EpiTxDb: creating an EpiTxDb object</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/FelixErnst/EpiTxDb/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="EpiTxDb_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>EpiTxDb: Storing and accessing epitranscriptomic information using the AnnotationDbi interface</h1>
                        <h4 class="author">Felix G.M. Ernst</h4>
            
            <h4 class="date">2021-01-09</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/FelixErnst/EpiTxDb/blob/master/vignettes/EpiTxDb.Rmd"><code>vignettes/EpiTxDb.Rmd</code></a></small>
      <div class="hidden name"><code>EpiTxDb.Rmd</code></div>

    </div>

    
    
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script><script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script><div id="installation" class="section level1">
<h1 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h1>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>

<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"EpiTxDb"</span>,<span class="st">"EpiTxDb.Hs.hg38"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>The epitranscriptome includes all post-transcriptional modifications of the RNA and describes and additional layer of information encoded on RNA. Like the term epigenome it is not about a change in nucleotide sequences, but the addition of functional elements through modifications.</p>
<p>With the development of high throughput detection strategies for specific RNA modifications, such as miCLIP and Pseudo-Seq amongst other, a large number of modified positions have been identified and were summarized via the RMBase project <a href="#References"><span class="citation">(Xuan et al. 2017; Sun et al. 2015)</span></a> project.</p>
<p>To make these information avaialble within the Bioconductor universe <code>EpiTxDb</code> was developed, which facilitates the storage of epitranscriptomic information. More specifically, it can keep track of modification identity, position, the enzyme for introducing it on the RNA, a specifier which determines the position on the RNA to be modified and the literature references each modification is associated with.</p>
</div>
<div id="getting-started" class="section level1">
<h1 class="hasAnchor">
<a href="#getting-started" class="anchor"></a>Getting started</h1>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/FelixErnst/EpiTxDb">EpiTxDb</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/FelixErnst/EpiTxDb.Hs.hg38">EpiTxDb.Hs.hg38</a></span><span class="op">)</span></code></pre></div>
<p>The class <code>EpiTxDb</code> is the class for storing the epitranscriptomic data. It inherits the inner workings of <code>AnnotationDb</code> class from the <code>AnnotationDbi</code> package.</p>
<p>As an example for the vignette the snoRNAdb data <a href="#References"><span class="citation">(Lestrade and Weber 2006)</span></a> from the <code>EpiTxDb.Hs.hg38</code> package will be used. The data is stored in the <code>AnnotationHub</code> and is downloaded and cached upon the first request.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">etdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/EpiTxDb.Hs.hg38/man/EpiTxDb.Hs.hg38.html">EpiTxDb.Hs.hg38.snoRNAdb</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## using temporary cache /tmp/RtmpDuN7ZR/BiocFileCache</code></pre>
<pre><code>## snapshotDate(): 2021-01-04</code></pre>
<pre><code>## downloading 1 resources</code></pre>
<pre><code>## retrieving 1 resource</code></pre>
<pre><code>## loading from cache</code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">etdb</span></code></pre></div>
<pre><code>## EpiTxDb object:
## # Db type: EpiTxDb
## # Supporting package: EpiTxDb
## # Data source: snoRNAdb
## # Organism: Homo sapiens
## # Genome: hg38
## # Coordinates: per Transcript
## # Nb of modifications: 235
## # Db created by: EpiTxDb package from Bioconductor
## # Creation time: 2020-02-26 10:34:30 +0100 (Wed, 26 Feb 2020)
## # EpiTxDb version at creation time: 0.99.0
## # RSQLite version at creation time: 2.2.0
## # DBSCHEMAVERSION: 1.0</code></pre>
<p>As expected for an <code>AnnotationDb</code> class the general accessors are available.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/select.html">keytypes</a></span><span class="op">(</span><span class="va">etdb</span><span class="op">)</span></code></pre></div>
<pre><code>##  [1] "MODID"            "MODNAME"          "MODSTRAND"        "MODTYPE"         
##  [5] "REF"              "REFTYPE"          "RXENSEMBL"        "RXENSEMBLTRANS"  
##  [9] "RXENTREZID"       "RXGENENAME"       "SNID"             "SNNAME"          
## [13] "SPECENSEMBL"      "SPECENSEMBLTRANS" "SPECENTREZID"     "SPECGENENAME"    
## [17] "SPECTYPE"</code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/select.html">columns</a></span><span class="op">(</span><span class="va">etdb</span><span class="op">)</span></code></pre></div>
<pre><code>##  [1] "MODEND"           "MODID"            "MODNAME"          "MODSTART"        
##  [5] "MODSTRAND"        "MODTYPE"          "REF"              "REFID"           
##  [9] "REFTYPE"          "RXENSEMBL"        "RXENSEMBLTRANS"   "RXENTREZID"      
## [13] "RXGENENAME"       "RXID"             "RXRANK"           "SNID"            
## [17] "SNNAME"           "SPECENSEMBL"      "SPECENSEMBLTRANS" "SPECENTREZID"    
## [21] "SPECGENENAME"     "SPECID"           "SPECTYPE"</code></pre>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/select.html">keys</a></span><span class="op">(</span><span class="va">etdb</span>, <span class="st">"MODID"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "1" "2" "3" "4" "5" "6"</code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/select.html">select</a></span><span class="op">(</span><span class="va">etdb</span>, keys <span class="op">=</span> <span class="st">"1"</span>,
       columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"MODNAME"</span>,<span class="st">"MODTYPE"</span>,<span class="st">"MODSTART"</span>,<span class="st">"MODSTRAND"</span>,<span class="st">"SNNAME"</span>,
                   <span class="st">"RXGENENAME"</span>,<span class="st">"SPECTYPE"</span>,<span class="st">"SPECGENENAME"</span><span class="op">)</span>,
       keytype <span class="op">=</span> <span class="st">"MODID"</span><span class="op">)</span></code></pre></div>
<pre><code>## 'select()' returned 1:1 mapping between keys and columns</code></pre>
<pre><code>##   MODID MODNAME MODTYPE MODSTART MODSTRAND    SNNAME  RXGENENAME SPECTYPE
## 1     1   Um_14      Um       14         + NR_003285 fibrillarin   snoRNA
##   SPECGENENAME
## 1      SNORD71</code></pre>
<p>The columns with the prefix <code>RX</code> or <code>SPEC</code> reference the reaction enzyme and the location specifier. This can be the same information, but for ribosomal modifications from the snoRNAdb it is of course fibrillarin and a snoRNA.</p>
<p>In addition the following accessor for metadata are available as well.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">species</span><span class="op">(</span><span class="va">etdb</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "Homo sapiens"</code></pre>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">organism</span><span class="op">(</span><span class="va">etdb</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "Homo sapiens"</code></pre>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">seqlevels</span><span class="op">(</span><span class="va">etdb</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "NR_003285" "NR_003286" "NR_003287" "NR_004430" "NR_002716" "NR_003925"
## [7] "NR_002756" "NR_004394" "NR_029422"</code></pre>
</div>
<div id="accessing-rna-modifications" class="section level1">
<h1 class="hasAnchor">
<a href="#accessing-rna-modifications" class="anchor"></a>Accessing RNA modifications</h1>
<p>The specialized accessors are <code><a href="../reference/modifications.html">modifications()</a></code> and <code><a href="../reference/modifications.html">modificationsBy()</a></code>. <code><a href="../reference/modifications.html">modifications()</a></code> allows for filtering results, whereas <code><a href="../reference/modifications.html">modificationsBy()</a></code> returns all the modifications in batches separated by certain information.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/modifications.html">modifications</a></span><span class="op">(</span><span class="va">etdb</span>, columns <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"mod_id"</span>,<span class="st">"mod_type"</span>,<span class="st">"mod_name"</span>,
                                <span class="st">"rx_genename"</span>,<span class="st">"spec_genename"</span>,
                                <span class="st">"ref_type"</span>,<span class="st">"ref"</span><span class="op">)</span>,
              filter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>mod_id <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## GRanges object with 3 ranges and 7 metadata columns:
##        seqnames    ranges strand |    mod_id    mod_type    mod_name
##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt; &lt;character&gt;
##   [1] NR_003285        14      + |         1          Um       Um_14
##   [2] NR_003285        55      + |         2           Y        Y_55
##   [3] NR_003285        69      + |         3           Y        Y_69
##                  rx_genename   spec_genename        ref_type             ref
##              &lt;CharacterList&gt; &lt;CharacterList&gt; &lt;CharacterList&gt; &lt;CharacterList&gt;
##   [1]            fibrillarin         SNORD71            PMID        16381836
##   [2] dyskerin pseudouridi..         SNORA72            PMID        16381836
##   [3] dyskerin pseudouridi..         SNORA69            PMID        16381836
##   -------
##   seqinfo: 9 sequences from hg38 genome; no seqlengths</code></pre>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># split by sequence name, usually a transcipt identifier</span>
<span class="fu"><a href="../reference/modifications.html">modificationsBy</a></span><span class="op">(</span><span class="va">etdb</span>, by <span class="op">=</span> <span class="st">"seqnames"</span><span class="op">)</span></code></pre></div>
<pre><code>## GRangesList object of length 9:
## $NR_003285
## GRanges object with 4 ranges and 3 metadata columns:
##        seqnames    ranges strand |    mod_id         mod    mod_name
##           &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt; &lt;character&gt;
##   [1] NR_003285        14      + |         1          Um       Um_14
##   [2] NR_003285        55      + |         2           Y        Y_55
##   [3] NR_003285        69      + |         3           Y        Y_69
##   [4] NR_003285        75      + |         4          Gm       Gm_75
##   -------
##   seqinfo: 9 sequences from hg38 genome; no seqlengths
## 
## ...
## &lt;8 more elements&gt;</code></pre>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># split modification type</span>
<span class="fu"><a href="../reference/modifications.html">modificationsBy</a></span><span class="op">(</span><span class="va">etdb</span>, by <span class="op">=</span> <span class="st">"modtype"</span><span class="op">)</span></code></pre></div>
<pre><code>## GRangesList object of length 5:
## $Am
## GRanges object with 39 ranges and 3 metadata columns:
##         seqnames    ranges strand |         mod    mod_id    mod_name
##            &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;integer&gt; &lt;character&gt;
##    [1] NR_003286        27      + |          Am         5       Am_27
##    [2] NR_003286        99      + |          Am         9       Am_99
##    [3] NR_003286       159      + |          Am        15      Am_159
##    [4] NR_003286       166      + |          Am        16      Am_166
##    [5] NR_003286       468      + |          Am        25      Am_468
##    ...       ...       ...    ... .         ...       ...         ...
##   [35] NR_002716        30      + |          Am       207       Am_30
##   [36] NR_003925        65      + |          Am       216       Am_65
##   [37] NR_004394        47      + |          Am       226       Am_47
##   [38] NR_004394        53      + |          Am       227       Am_53
##   [39] NR_004394        53      + |          Am       228       Am_53
##   -------
##   seqinfo: 9 sequences from hg38 genome; no seqlengths
## 
## ...
## &lt;4 more elements&gt;</code></pre>
</div>
<div id="shifting-coordinates-from-genomic-to-transcritomic" class="section level1">
<h1 class="hasAnchor">
<a href="#shifting-coordinates-from-genomic-to-transcritomic" class="anchor"></a>Shifting coordinates from genomic to transcritomic</h1>
<p>Since epitranscriptomic modifications by their nature refere to each individual it might be of interest to shift the coordinates transctipt coordinates keeping in mind that with transcript variants multiple options exist for each of the transcript maturation process.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">TxDb.Hsapiens.UCSC.hg38.knownGene</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">BSgenome.Hsapiens.UCSC.hg38</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">txdb</span> <span class="op">&lt;-</span> <span class="va">TxDb.Hsapiens.UCSC.hg38.knownGene</span>
<span class="fu">seqlevels</span><span class="op">(</span><span class="va">txdb</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"chr1"</span>
<span class="va">bs</span> <span class="op">&lt;-</span> <span class="va">BSgenome.Hsapiens.UCSC.hg38</span>

<span class="va">etdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/EpiTxDb.Hs.hg38/man/EpiTxDb.Hs.hg38.html">EpiTxDb.Hs.hg38.RMBase</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## snapshotDate(): 2021-01-04</code></pre>
<pre><code>## downloading 1 resources</code></pre>
<pre><code>## retrieving 1 resource</code></pre>
<pre><code>## loading from cache</code></pre>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">tx</span> <span class="op">&lt;-</span> <span class="fu">exonsBy</span><span class="op">(</span><span class="va">txdb</span><span class="op">)</span>
<span class="va">mod</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/modifications.html">modifications</a></span><span class="op">(</span><span class="va">etdb</span>, filter <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>sn_name <span class="op">=</span> <span class="st">"chr1"</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">mod</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 47275</code></pre>
<p>In the following example we will focus on shifting the coordinates to individual mature transcripts. However, keep in mind, that premature transcript might be of interest as well and this can be controlled via the <code>tx</code> arguments of <code><a href="../reference/shiftTranscriptToGenomic.html">shiftGenomicToTranscript()</a></code></p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mod_tx</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/shiftTranscriptToGenomic.html">shiftGenomicToTranscript</a></span><span class="op">(</span><span class="va">mod</span>, <span class="va">tx</span><span class="op">)</span></code></pre></div>
<pre><code>## Warning: Coordinates for 905 ranges of 'subject' not found:
## 'chr1:14662:-','chr1:14668:-','chr1:14766:-','chr1:139005:-','chr1:139020:-','chr1:139040:-','chr1:139127:-','chr1:629454:+','chr1:629456:+','chr1:629465:+'
## and more ...</code></pre>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">mod_tx</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 157694</code></pre>
<p>Due to multiple matches for each transcript variant the number of modifications has increased.</p>
<p>With the we can plot the relative positions of modifications by type on <code>chr1</code> transcripts.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">mod_tx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/split.html">split</a></span><span class="op">(</span><span class="va">mod_tx</span>,<span class="fu">seqnames</span><span class="op">(</span><span class="va">mod_tx</span><span class="op">)</span><span class="op">)</span>
<span class="va">names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/funprog.html">Reduce</a></span><span class="op">(</span><span class="va">intersect</span>,<span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">mod_tx</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">tx</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="co"># Getting the corresponding 5'-UTR and 3'-UTR annotations</span>
<span class="va">fp</span> <span class="op">&lt;-</span> <span class="fu">fiveUTRsByTranscript</span><span class="op">(</span><span class="va">txdb</span><span class="op">)</span>
<span class="va">tp</span> <span class="op">&lt;-</span> <span class="fu">threeUTRsByTranscript</span><span class="op">(</span><span class="va">txdb</span><span class="op">)</span>
<span class="va">tx</span> <span class="op">&lt;-</span> <span class="va">tx</span><span class="op">[</span><span class="va">names</span><span class="op">]</span>
<span class="va">mod_tx</span> <span class="op">&lt;-</span> <span class="va">mod_tx</span><span class="op">[</span><span class="va">names</span><span class="op">]</span>
<span class="va">fp_m</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span><span class="op">(</span><span class="va">names</span>,<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">fp</span><span class="op">)</span><span class="op">)</span>
<span class="va">fp_m</span> <span class="op">&lt;-</span> <span class="va">fp_m</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">fp_m</span><span class="op">)</span><span class="op">]</span>
<span class="va">tp_m</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/match.html">match</a></span><span class="op">(</span><span class="va">names</span>,<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">tp</span><span class="op">)</span><span class="op">)</span>
<span class="va">tp_m</span> <span class="op">&lt;-</span> <span class="va">tp_m</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">tp_m</span><span class="op">)</span><span class="op">]</span>
<span class="va">fp</span> <span class="op">&lt;-</span> <span class="va">fp</span><span class="op">[</span><span class="va">fp_m</span><span class="op">]</span>
<span class="va">tp</span> <span class="op">&lt;-</span> <span class="va">tp</span><span class="op">[</span><span class="va">tp_m</span><span class="op">]</span>

<span class="co"># Getting lengths of transcripts, 5'-UTR and 3'-UTR</span>
<span class="va">tx_lengths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="fu">width</span><span class="op">(</span><span class="va">tx</span><span class="op">)</span><span class="op">)</span>
<span class="va">fp_lengths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">0L</span>,<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">tx</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">fp_lengths</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">names</span>
<span class="va">fp_lengths</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">fp</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="fu">width</span><span class="op">(</span><span class="va">fp</span><span class="op">)</span><span class="op">)</span>
<span class="va">tp_lengths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html">rep</a></span><span class="op">(</span><span class="fl">0L</span>,<span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">tx</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">tp_lengths</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">names</span>
<span class="va">tp_lengths</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">tp</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="fu">width</span><span class="op">(</span><span class="va">tp</span><span class="op">)</span><span class="op">)</span>

<span class="co"># Rescale modifications</span>
<span class="co"># CDS start is at position 1L and cds end at position 1000L</span>
<span class="va">from</span> <span class="op">&lt;-</span> <span class="fu">IRanges</span><span class="op">(</span><span class="va">fp_lengths</span><span class="op">+</span><span class="fl">1L</span>, <span class="va">tx_lengths</span> <span class="op">-</span> <span class="va">tp_lengths</span><span class="op">)</span>
<span class="va">to</span> <span class="op">&lt;-</span> <span class="fu">IRanges</span><span class="op">(</span><span class="fl">1L</span>,<span class="fl">1000L</span><span class="op">)</span>
<span class="va">mod_rescale</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/rescale.html">rescale</a></span><span class="op">(</span><span class="va">mod_tx</span>, <span class="va">to</span>, <span class="va">from</span><span class="op">)</span>

<span class="co"># Construct result data.frame</span>
<span class="va">rel_pos</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>mod_type <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="fu">mcols</span><span class="op">(</span><span class="va">mod_rescale</span>,level<span class="op">=</span><span class="st">"within"</span><span class="op">)</span><span class="op">[</span>,<span class="st">"mod_type"</span><span class="op">]</span><span class="op">)</span>,
                      rel_pos <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/start.html">start</a></span><span class="op">(</span><span class="va">mod_rescale</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="va">rel_pos</span> <span class="op">&lt;-</span> <span class="va">rel_pos</span><span class="op">[</span><span class="va">rel_pos</span><span class="op">$</span><span class="va">rel_pos</span> <span class="op">&lt;</span> <span class="fl">1500</span> <span class="op">&amp;</span> <span class="va">rel_pos</span><span class="op">$</span><span class="va">rel_pos</span> <span class="op">&gt;</span> <span class="op">-</span><span class="fl">500</span>,<span class="op">]</span></code></pre></div>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>
<span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">rel_pos</span><span class="op">[</span><span class="va">rel_pos</span><span class="op">$</span><span class="va">mod_type</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"m6A"</span>,<span class="st">"m1A"</span>,<span class="st">"Y"</span><span class="op">)</span>,<span class="op">]</span>,
       <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">rel_pos</span>, colour <span class="op">=</span> <span class="va">mod_type</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="EpiTxDb_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
</div>
<div id="session-info" class="section level1">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session info</h1>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## R Under development (unstable) (2021-01-07 r79806)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 20.04.1 LTS
## 
## Matrix products: default
## BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] ggplot2_3.3.3                           
##  [2] BSgenome.Hsapiens.UCSC.hg38_1.4.3       
##  [3] BSgenome_1.59.2                         
##  [4] rtracklayer_1.51.3                      
##  [5] TxDb.Hsapiens.UCSC.hg38.knownGene_3.10.0
##  [6] GenomicFeatures_1.43.3                  
##  [7] GenomicRanges_1.43.1                    
##  [8] GenomeInfoDb_1.27.3                     
##  [9] EpiTxDb.Hs.hg38_0.99.7                  
## [10] AnnotationHub_2.23.0                    
## [11] BiocFileCache_1.15.1                    
## [12] dbplyr_2.0.0                            
## [13] EpiTxDb_1.3.1                           
## [14] Modstrings_1.7.1                        
## [15] Biostrings_2.59.2                       
## [16] XVector_0.31.1                          
## [17] AnnotationDbi_1.53.0                    
## [18] IRanges_2.25.6                          
## [19] S4Vectors_0.29.6                        
## [20] Biobase_2.51.0                          
## [21] BiocGenerics_0.37.0                     
## [22] BiocStyle_2.19.1                        
## 
## loaded via a namespace (and not attached):
##  [1] bitops_1.0-6                  matrixStats_0.57.0           
##  [3] fs_1.5.0                      bit64_4.0.5                  
##  [5] filelock_1.0.2                progress_1.2.2               
##  [7] httr_1.4.2                    rprojroot_2.0.2              
##  [9] tools_4.1.0                   R6_2.5.0                     
## [11] DBI_1.1.0                     colorspace_2.0-0             
## [13] withr_2.3.0                   tidyselect_1.1.0             
## [15] prettyunits_1.1.1             bit_4.0.4                    
## [17] curl_4.3                      compiler_4.1.0               
## [19] textshaping_0.2.1             xml2_1.3.2                   
## [21] desc_1.2.0                    DelayedArray_0.17.7          
## [23] labeling_0.4.2                bookdown_0.21                
## [25] scales_1.1.1                  tRNA_1.9.0                   
## [27] askpass_1.1                   rappdirs_0.3.1               
## [29] pkgdown_1.6.1                 systemfonts_0.3.2            
## [31] stringr_1.4.0                 digest_0.6.27                
## [33] Rsamtools_2.7.0               rmarkdown_2.6                
## [35] pkgconfig_2.0.3               htmltools_0.5.0              
## [37] MatrixGenerics_1.3.0          fastmap_1.0.1                
## [39] rlang_0.4.10                  RSQLite_2.2.2                
## [41] shiny_1.5.0                   farver_2.0.3                 
## [43] BiocIO_1.1.2                  generics_0.1.0               
## [45] BiocParallel_1.25.2           dplyr_1.0.2                  
## [47] RCurl_1.98-1.2                magrittr_2.0.1               
## [49] GenomeInfoDbData_1.2.4        Matrix_1.3-2                 
## [51] Rcpp_1.0.5                    munsell_0.5.0                
## [53] lifecycle_0.2.0               stringi_1.5.3                
## [55] yaml_2.2.1                    SummarizedExperiment_1.21.1  
## [57] zlibbioc_1.37.0               grid_4.1.0                   
## [59] blob_1.2.1                    promises_1.1.1               
## [61] crayon_1.3.4                  lattice_0.20-41              
## [63] tRNAdbImport_1.9.0            hms_0.5.3                    
## [65] knitr_1.30                    pillar_1.4.7                 
## [67] rjson_0.2.20                  biomaRt_2.47.1               
## [69] BiocVersion_3.13.1            XML_3.99-0.5                 
## [71] glue_1.4.2                    evaluate_0.14                
## [73] Structstrings_1.7.2           BiocManager_1.30.10          
## [75] httpuv_1.5.4                  vctrs_0.3.6                  
## [77] gtable_0.3.0                  openssl_1.4.3                
## [79] purrr_0.3.4                   assertthat_0.2.1             
## [81] xfun_0.20                     mime_0.9                     
## [83] xtable_1.8-4                  restfulr_0.0.13              
## [85] later_1.1.0.1                 ragg_0.4.0                   
## [87] tibble_3.0.4                  GenomicAlignments_1.27.2     
## [89] memoise_1.1.0                 interactiveDisplayBase_1.29.0
## [91] ellipsis_0.3.1</code></pre>
<p><a name="References"></a></p>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-Lestrade-2006">
<p>Lestrade, Laurent, and Michel J. Weber. 2006. “snoRNA-LBME-db, a comprehensive database of human H/ACA and C/D box snoRNAs.” <em>Nucleic Acids Research</em> 34 (January): D158–D162. <a href="https://doi.org/10.1093/nar/gkj002">https://doi.org/10.1093/nar/gkj002</a>.</p>
</div>
<div id="ref-Sun-2015">
<p>Sun, Wen-Ju, Jun-Hao Li, Shun Liu, Jie Wu, Hui Zhou, Liang-Hu Qu, and Jian-Hua Yang. 2015. “RMBase: a resource for decoding the landscape of RNA modifications from high-throughput sequencing data.” <em>Nucleic Acids Research</em> 44 (D1): D259–D265. <a href="https://doi.org/10.1093/nar/gkv1036">https://doi.org/10.1093/nar/gkv1036</a>.</p>
</div>
<div id="ref-Xuan-2017">
<p>Xuan, Jia-Jia, Wen-Ju Sun, Peng-Hui Lin, Ke-Ren Zhou, Shun Liu, Ling-Ling Zheng, Liang-Hu Qu, and Jian-Hua Yang. 2017. “RMBase v2.0: deciphering the map of RNA modifications from epitranscriptome sequencing data.” <em>Nucleic Acids Research</em> 46 (D1): D327–D334. <a href="https://doi.org/10.1093/nar/gkx934">https://doi.org/10.1093/nar/gkx934</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Felix G.M. Ernst.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
