<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>EpiTxDb: creating an EpiTxDb object • EpiTxDb</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="EpiTxDb: creating an EpiTxDb object">
<meta property="og:description" content="EpiTxDb">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">EpiTxDb</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.3.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/EpiTxDb.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/EpiTxDb-creation.html">EpiTxDb: creating an EpiTxDb object</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/FelixErnst/EpiTxDb/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="EpiTxDb-creation_files/accessible-code-block-0.0.1/empty-anchor.js"></script><link href="EpiTxDb-creation_files/anchor-sections-1.0/anchor-sections.css" rel="stylesheet">
<script src="EpiTxDb-creation_files/anchor-sections-1.0/anchor-sections.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>EpiTxDb: creating an EpiTxDb object</h1>
                        <h4 class="author">Felix G.M. Ernst</h4>
            
            <h4 class="date">2020-12-13</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/FelixErnst/EpiTxDb/blob/master/vignettes/EpiTxDb-creation.Rmd"><code>vignettes/EpiTxDb-creation.Rmd</code></a></small>
      <div class="hidden name"><code>EpiTxDb-creation.Rmd</code></div>

    </div>

    
    
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script><script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script><div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>To create an <code>EpiTxDb</code> object a number of different functions are available. The most univeral functions are <code>makeEpiTxDb</code> and <code>makeEpiTxDbFromGRanges</code>. <code>makeEpiTxDb</code> uses four <code>data.frame</code>s as input, whereas <code>makeEpiTxDbFromGRanges</code> is a wrapper for information available as a <code>GRanges</code> object.</p>
<p>The other functions are <code>makeEpiTxDbFromRMBase</code> and <code>makeEpiTxDbFromtRNAdb</code>, which are aimed to make data available from the RMBase v2.0 database <a href="#References"><span class="citation">(Xuan et al. 2017; Sun et al. 2015)</span></a> or the tRNAdb <a href="#References"><span class="citation">(Jühling et al. 2009; Sprinzl and Vassilenko 2005)</span></a>. However, before creating your <code>EpiTxDb</code> objects, have a look at the already available resources for <code>H. sapiens</code>, <code>M. musculus</code> and <code>S. cerevisiae.</code></p>
<p>Additional metadata can be provided as separate <code>data.frame</code> for all functions. The <code>data.frame</code> must have two columns <code>name</code> and <code>value</code>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/GenomicRanges">GenomicRanges</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/FelixErnst/EpiTxDb">EpiTxDb</a></span><span class="op">)</span></code></pre></div>
</div>
<div id="makeepitxdb-and-makeepitxdbfromgranges" class="section level1">
<h1 class="hasAnchor">
<a href="#makeepitxdb-and-makeepitxdbfromgranges" class="anchor"></a><code>makeEpiTxDb</code> and <code>makeEpiTxDbFromGRanges</code>
</h1>
<p>The creation of an etdb object is quite easy starting with a <code>GRanges</code> object.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">gr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html">GRanges</a></span><span class="op">(</span>seqnames <span class="op">=</span> <span class="st">"test"</span>,
              ranges <span class="op">=</span> <span class="fu">IRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html">IRanges</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">1</span><span class="op">)</span>,
              strand <span class="op">=</span> <span class="st">"+"</span>,
              <span class="fu">DataFrame</span><span class="op">(</span>mod_id <span class="op">=</span> <span class="fl">1L</span>,
                        mod_type <span class="op">=</span> <span class="st">"Am"</span>,
                        mod_name <span class="op">=</span> <span class="st">"Am_1"</span><span class="op">)</span><span class="op">)</span>
<span class="va">etdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/makeEpiTxDbFromGRanges.html">makeEpiTxDbFromGRanges</a></span><span class="op">(</span><span class="va">gr</span>, metadata <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>name <span class="op">=</span> <span class="st">"test"</span>,
                                                         value <span class="op">=</span> <span class="st">"Yes"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## Creating EpiTxDb object ... done</code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">etdb</span></code></pre></div>
<pre><code>## EpiTxDb object:
## # Db type: EpiTxDb
## # Supporting package: EpiTxDb
## # test: Yes
## # Nb of modifications: 1
## # Db created by: EpiTxDb package from Bioconductor
## # Creation time: 2020-12-13 13:00:18 +0000 (Sun, 13 Dec 2020)
## # EpiTxDb version at creation time: 1.3.1
## # RSQLite version at creation time: 2.2.1
## # DBSCHEMAVERSION: 1.0</code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">metadata</span><span class="op">(</span><span class="va">etdb</span><span class="op">)</span></code></pre></div>
<pre><code>##                               name                                        value
## 1                          Db type                                      EpiTxDb
## 2               Supporting package                                      EpiTxDb
## 3                             test                                          Yes
## 4              Nb of modifications                                            1
## 5                    Db created by            EpiTxDb package from Bioconductor
## 6                    Creation time 2020-12-13 13:00:18 +0000 (Sun, 13 Dec 2020)
## 7 EpiTxDb version at creation time                                        1.3.1
## 8 RSQLite version at creation time                                        2.2.1
## 9                  DBSCHEMAVERSION                                          1.0</code></pre>
<p>Additional data can be provided via the metadata columns of the <code>GRanges</code> object. For supported columns have a look at <code><a href="../reference/makeEpiTxDb.html">?makeEpiTxDb</a></code> or <code><a href="../reference/makeEpiTxDbFromGRanges.html">?makeEpiTxDbFromGRanges</a></code>.</p>
</div>
<div id="makeepitxdbfromtrnadb" class="section level1">
<h1 class="hasAnchor">
<a href="#makeepitxdbfromtrnadb" class="anchor"></a><code>makeEpiTxDbFromtRNAdb</code>
</h1>
<p>The information of the tRNAdb can be accessed via the <code>tRNAdbImport</code> package using the RNA database. As a result a <code>ModRNAStringSet</code> object is returned from which the modifications can be extracted using <code>separate()</code>.</p>
<p>The only input require is a valid organism name returned by <code><a href="../reference/makeEpiTxDbFromtRNAdb.html">listAvailableOrganismsFromtRNAdb()</a></code>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">etdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/makeEpiTxDbFromtRNAdb.html">makeEpiTxDbFromtRNAdb</a></span><span class="op">(</span><span class="st">"Saccharomyces cerevisiae"</span><span class="op">)</span></code></pre></div>
<pre><code>## Loading data from tRNAdb ...</code></pre>
<pre><code>## Assembling data ...</code></pre>
<pre><code>## Creating EpiTxDb object ... done</code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">etdb</span></code></pre></div>
<pre><code>## EpiTxDb object:
## # Db type: EpiTxDb
## # Supporting package: EpiTxDb
## # Nb of modifications: 557
## # Db created by: EpiTxDb package from Bioconductor
## # Creation time: 2020-12-13 13:00:46 +0000 (Sun, 13 Dec 2020)
## # EpiTxDb version at creation time: 1.3.1
## # RSQLite version at creation time: 2.2.1
## # DBSCHEMAVERSION: 1.0</code></pre>
<p>For additional information have a look at <code><a href="../reference/makeEpiTxDbFromtRNAdb.html">?makeEpiTxDbFromtRNAdb</a></code>. The result returned from the tRNAdb is also available as <code>GRanges</code> object, if <code><a href="../reference/makeEpiTxDbFromtRNAdb.html">gettRNAdbDataAsGRanges()</a></code> is used.</p>
</div>
<div id="makeepitxdbfromrmbase" class="section level1">
<h1 class="hasAnchor">
<a href="#makeepitxdbfromrmbase" class="anchor"></a><code>makeEpiTxDbFromRMBase</code>
</h1>
<p>Analogous to the example above <code><a href="../reference/makeEpiTxDbFromRMBase.html">makeEpiTxDbFromRMBase()</a></code> will download the data from the RMBase v2.0. Three inputs are required, <code>organism</code>, <code>genome</code> and <code>modtype</code>, which have to valid bia the functions <code><a href="../reference/makeEpiTxDbFromRMBase.html">listAvailableOrganismsFromRMBase()</a></code>, <code>.listAvailableGenomesFromRMBase()</code> and <code>listAvailableModFromRMBase</code>.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">etdb</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/makeEpiTxDbFromRMBase.html">makeEpiTxDbFromRMBase</a></span><span class="op">(</span>organism <span class="op">=</span> <span class="st">"Saccharomyces cerevisiae"</span>,
                              genome <span class="op">=</span> <span class="st">"sacCer3"</span>,
                              modtype <span class="op">=</span> <span class="st">"m1A"</span><span class="op">)</span></code></pre></div>
<p>Internally, the files are cached using the <code>BiocFileCache</code> package and passed to <code><a href="../reference/makeEpiTxDbFromRMBase.html">makeEpiTxDbFromRMBaseFiles()</a></code>, which can also be used with locally stored files. The resuls for creating the <code>EpiTxDb</code> class are processed from these files via the <code><a href="../reference/makeEpiTxDbFromRMBase.html">getRMBaseDataAsGRanges()</a></code> function.</p>
</div>
<div id="session-info" class="section level1">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session info</h1>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## R Under development (unstable) (2020-11-02 r79396)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 20.04.1 LTS
## 
## Matrix products: default
## BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] EpiTxDb_1.3.1        Modstrings_1.7.0     Biostrings_2.59.0   
##  [4] XVector_0.31.0       AnnotationDbi_1.53.0 Biobase_2.51.0      
##  [7] GenomicRanges_1.43.0 GenomeInfoDb_1.27.2  IRanges_2.25.4      
## [10] S4Vectors_0.29.6     BiocGenerics_0.37.0  BiocStyle_2.19.1    
## 
## loaded via a namespace (and not attached):
##  [1] MatrixGenerics_1.3.0        httr_1.4.2                 
##  [3] bit64_4.0.5                 assertthat_0.2.1           
##  [5] askpass_1.1                 BiocManager_1.30.10        
##  [7] BiocFileCache_1.15.1        blob_1.2.1                 
##  [9] GenomeInfoDbData_1.2.4      Rsamtools_2.7.0            
## [11] yaml_2.2.1                  progress_1.2.2             
## [13] lattice_0.20-41             pillar_1.4.7               
## [15] RSQLite_2.2.1               glue_1.4.2                 
## [17] digest_0.6.27               Structstrings_1.7.1        
## [19] colorspace_2.0-0            tRNA_1.9.0                 
## [21] Matrix_1.2-18               htmltools_0.5.0            
## [23] XML_3.99-0.5                pkgconfig_2.0.3            
## [25] biomaRt_2.47.1              bookdown_0.21              
## [27] zlibbioc_1.37.0             purrr_0.3.4                
## [29] scales_1.1.1                BiocParallel_1.25.1        
## [31] tibble_3.0.4                openssl_1.4.3              
## [33] ggplot2_3.3.2               generics_0.1.0             
## [35] ellipsis_0.3.1              SummarizedExperiment_1.21.0
## [37] GenomicFeatures_1.43.2      magrittr_2.0.1             
## [39] crayon_1.3.4                memoise_1.1.0              
## [41] evaluate_0.14               fs_1.5.0                   
## [43] xml2_1.3.2                  textshaping_0.2.1          
## [45] tools_4.1.0                 prettyunits_1.1.1          
## [47] hms_0.5.3                   matrixStats_0.57.0         
## [49] BiocIO_1.1.2                lifecycle_0.2.0            
## [51] stringr_1.4.0               munsell_0.5.0              
## [53] tRNAdbImport_1.9.0          DelayedArray_0.17.4        
## [55] compiler_4.1.0              pkgdown_1.6.1              
## [57] systemfonts_0.3.2           rlang_0.4.9                
## [59] grid_4.1.0                  RCurl_1.98-1.2             
## [61] rappdirs_0.3.1              bitops_1.0-6               
## [63] rmarkdown_2.5               gtable_0.3.0               
## [65] DBI_1.1.0                   curl_4.3                   
## [67] R6_2.5.0                    GenomicAlignments_1.27.1   
## [69] knitr_1.30                  dplyr_1.0.2                
## [71] rtracklayer_1.51.1          bit_4.0.4                  
## [73] filelock_1.0.2              rprojroot_2.0.2            
## [75] ragg_0.4.0                  desc_1.2.0                 
## [77] stringi_1.5.3               Rcpp_1.0.5                 
## [79] vctrs_0.3.5                 dbplyr_2.0.0               
## [81] tidyselect_1.1.0            xfun_0.19</code></pre>
<p><a name="References"></a></p>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-Juehling-2009">
<p>Jühling, Frank, Mario Mörl, Roland K. Hartmann, Mathias Sprinzl, Peter F. Stadler, and Joern Pütz. 2009. “TRNAdb 2009: Compilation of tRNA Sequences and tRNA Genes.” <em>Nucleic Acids Research</em> 37: D159–D162. <a href="https://doi.org/10.1093/nar/gkn772">https://doi.org/10.1093/nar/gkn772</a>.</p>
</div>
<div id="ref-Sprinzl-2005">
<p>Sprinzl, Mathias, and Konstantin S. Vassilenko. 2005. “Compilation of tRNA Sequences and Sequences of tRNA Genes.” <em>Nucleic Acids Research</em> 33: D139–D140. <a href="https://doi.org/10.1093/nar/gki012">https://doi.org/10.1093/nar/gki012</a>.</p>
</div>
<div id="ref-Sun-2015">
<p>Sun, Wen-Ju, Jun-Hao Li, Shun Liu, Jie Wu, Hui Zhou, Liang-Hu Qu, and Jian-Hua Yang. 2015. “RMBase: a resource for decoding the landscape of RNA modifications from high-throughput sequencing data.” <em>Nucleic Acids Research</em> 44 (D1): D259–D265. <a href="https://doi.org/10.1093/nar/gkv1036">https://doi.org/10.1093/nar/gkv1036</a>.</p>
</div>
<div id="ref-Xuan-2017">
<p>Xuan, Jia-Jia, Wen-Ju Sun, Peng-Hui Lin, Ke-Ren Zhou, Shun Liu, Ling-Ling Zheng, Liang-Hu Qu, and Jian-Hua Yang. 2017. “RMBase v2.0: deciphering the map of RNA modifications from epitranscriptome sequencing data.” <em>Nucleic Acids Research</em> 46 (D1): D327–D334. <a href="https://doi.org/10.1093/nar/gkx934">https://doi.org/10.1093/nar/gkx934</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Felix G.M. Ernst.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
